File: go/swagger20_model.go
package openapiclient

import (
	"context"
	"encoding/json"
	"net/http"
)

// Swagger20Source identifies an OpenAPI 2.0 (Swagger) artifact without an
// OpenBindings document. Content is the artifact's JSON or YAML source image;
// a co-present Location supplies its canonical reference base.
type Swagger20Source struct {
	Location string
	Content  []byte
	Document *Swagger20Document
}

// Swagger20Document is the native, raw-preserving OpenAPI 2.0 document model.
// Its representation stays private so callers cannot mutate a prepared
// reference graph; MarshalJSON exposes a lossless JSON image when inspection
// is needed.
type Swagger20Document struct {
	root  swagger20Object
	entry *swagger20Resource
	graph *swagger20ReferenceGraph
}

func (d *Swagger20Document) rebind(ctx context.Context, client *http.Client, allowExternalRefs bool) *Swagger20Document {
	graph := newSwagger20ReferenceGraph(ctx, client, d.graph.selfContained, allowExternalRefs)
	entry := graph.rememberResolvedResource(d.entry.requested, d.entry.retrieval, d.root)
	return &Swagger20Document{root: d.root, entry: entry, graph: graph}
}

func (d *Swagger20Document) MarshalJSON() ([]byte, error) {
	if d == nil {
		return []byte("null"), nil
	}
	return json.Marshal(d.root)
}

// Swagger reports the exact root discriminator retained by the native model.
func (d *Swagger20Document) Swagger() string {
	if d == nil {
		return ""
	}
	return d.root.string("swagger").value
}

// Presence-aware members keep absence, explicit empty values, wrong JSON
// types, and valid values distinct. The 2.0 execution and synthesis lanes use
// these accessors instead of zero-valued Go struct fields.
type swagger20Member[T any] struct {
	value   T
	present bool
	valid   bool
}

type swagger20Object map[string]any

func (o swagger20Object) member(name string) (any, bool) {
	if o == nil {
		return nil, false
	}
	value, present := o[name]
	return value, present
}

func (o swagger20Object) string(name string) swagger20Member[string] {
	value, present := o.member(name)
	if !present {
		return swagger20Member[string]{}
	}
	text, valid := value.(string)
	return swagger20Member[string]{value: text, present: true, valid: valid}
}

func (o swagger20Object) boolean(name string) swagger20Member[bool] {
	value, present := o.member(name)
	if !present {
		return swagger20Member[bool]{}
	}
	boolean, valid := value.(bool)
	return swagger20Member[bool]{value: boolean, present: true, valid: valid}
}

func (o swagger20Object) object(name string) swagger20Member[swagger20Object] {
	value, present := o.member(name)
	if !present {
		return swagger20Member[swagger20Object]{}
	}
	object, valid := value.(map[string]any)
	return swagger20Member[swagger20Object]{value: swagger20Object(object), present: true, valid: valid}
}

func (o swagger20Object) array(name string) swagger20Member[[]any] {
	value, present := o.member(name)
	if !present {
		return swagger20Member[[]any]{}
	}
	array, valid := value.([]any)
	return swagger20Member[[]any]{value: array, present: true, valid: valid}
}

type swagger20DocumentObject struct {
	raw      swagger20Object
	resource *swagger20Resource
}

func (d swagger20DocumentObject) paths() swagger20Member[swagger20Object] {
	return d.raw.object("paths")
}

type swagger20PathItem struct {
	raw             swagger20Object
	resource        *swagger20Resource
	memberResources map[string]*swagger20Resource
}

func (p swagger20PathItem) reference() swagger20Member[string] {
	return p.raw.string("$ref")
}

func (p swagger20PathItem) operation(method string) swagger20Member[swagger20Object] {
	return p.raw.object(method)
}

func (p swagger20PathItem) parameters() swagger20Member[[]any] {
	return p.raw.array("parameters")
}

func (p swagger20PathItem) resourceFor(member string) *swagger20Resource {
	if resource := p.memberResources[member]; resource != nil {
		return resource
	}
	return p.resource
}

type swagger20Operation struct {
	raw      swagger20Object
	resource *swagger20Resource
	pathItem swagger20PathItem
	path     string
	method   string
}

func (o swagger20Operation) operationInfo(ref string) OperationInfo {
	operationID := o.raw.string("operationId")
	summary := o.raw.string("summary")
	info := OperationInfo{
		Ref: ref, Path: o.path, Method: Method(o.method),
		OperationID: operationID.value, Summary: summary.value,
	}
	if tags := o.raw.array("tags"); tags.present && tags.valid {
		for _, value := range tags.value {
			if tag, ok := value.(string); ok {
				info.Tags = append(info.Tags, tag)
			}
		}
	}
	return info
}

func swagger20Method(method string) bool {
	switch method {
	case "get", "put", "post", "delete", "options", "head", "patch":
		return true
	default:
		return false
	}
}

func swagger20HTTPMethod(method string) string {
	switch method {
	case "get":
		return http.MethodGet
	case "put":
		return http.MethodPut
	case "post":
		return http.MethodPost
	case "delete":
		return http.MethodDelete
	case "options":
		return http.MethodOptions
	case "head":
		return http.MethodHead
	case "patch":
		return http.MethodPatch
	default:
		return ""
	}
}

File: typescript/src/swagger20-model.ts
/** A JSON object retained exactly as authored by a Swagger 2.0 resource. */
export type Swagger20Object = Record<string, unknown>;

/** OpenAPI-native source carriage for an exact Swagger 2.0 artifact. */
export interface Swagger20Source {
  /** Canonical document URI and retrieval address when content is absent. */
  location?: string;
  /** Parsed JSON object, JSON/YAML text, or UTF-8 bytes. Content has primacy. */
  content?: unknown;
  /** A document previously returned by loadSwagger20. */
  document?: Swagger20Document;
}

export interface Swagger20LoadOptions {
  fetch?: typeof globalThis.fetch;
  signal?: AbortSignal;
  /** Defaults to true. False keeps inspection strictly side-effect-free. */
  allowExternalRefs?: boolean;
}

/** One native Swagger 2.0 operation identity. */
export interface Swagger20OperationInfo {
  ref: string;
  path: string;
  method: Swagger20Method;
  operationId?: string;
  summary?: string;
  tags: string[];
}

export type Swagger20Method = "get" | "put" | "post" | "delete" | "options" | "head" | "patch";

/**
 * Exact JSON-number carrier for hosts that preserve the source token. Plain
 * JavaScript numbers remain accepted, but this form keeps distinctions such
 * as `1` versus `1.0` available to draft-04's literal-integer rule.
 */
export class Swagger20Number {
  readonly lexeme: string;
  readonly value: number;

  constructor(lexeme: string) {
    if (!/^-?(?:0|[1-9][0-9]*)(?:\.[0-9]+)?(?:[eE][+-]?[0-9]+)?$/u.test(lexeme)) {
      throw new Error(`invalid JSON number token ${JSON.stringify(lexeme)}`);
    }
    const value = Number(lexeme);
    if (!Number.isFinite(value)) throw new Error(`JSON number token ${JSON.stringify(lexeme)} is not finite`);
    this.lexeme = lexeme;
    this.value = value;
  }

  toJSON(): number {
    return this.value;
  }

  toString(): string {
    return this.lexeme;
  }
}

export const SWAGGER20_METHODS: readonly Swagger20Method[] =
  Object.freeze(["get", "put", "post", "delete", "options", "head", "patch"]);

/** @internal */
export interface Swagger20Resource {
  requested?: string;
  retrieval?: string;
  root: unknown;
}

/** @internal */
export interface Swagger20ResolvedOperation {
  raw: Swagger20Object;
  resource: Swagger20Resource;
  graph: Swagger20ReferenceGraphContract;
  pathItem: Swagger20PathItem;
  path: string;
  method: Swagger20Method;
}

/** @internal */
export interface Swagger20PathItem {
  raw: Swagger20Object;
  resource: Swagger20Resource;
  memberResources?: Map<string, Swagger20Resource>;
}

/**
 * Native, raw-preserving Swagger 2.0 document model. It is deliberately not
 * assignable to the package's OpenAPI 3.x structural model.
 */
export class Swagger20Document {
  /** @internal */ readonly root: Swagger20Object;
  /** @internal */ readonly entry: Swagger20Resource;
  /** @internal */ readonly graph: Swagger20ReferenceGraphContract;

  /** @internal */
  constructor(root: Swagger20Object, entry: Swagger20Resource, graph: Swagger20ReferenceGraphContract) {
    this.root = root;
    this.entry = entry;
    this.graph = graph;
  }

  /** Exact retained root discriminator. */
  get swagger(): string | undefined {
    return stringMember(this.root, "swagger").value;
  }

  /** Lossless JSON-domain inspection image. */
  toJSON(): Swagger20Object {
    return this.root;
  }
}

/** @internal - breaks the model/reference module cycle without widening the public surface. */
export interface Swagger20ReferenceGraphContract {
  readonly selfContained: boolean;
  readonly allowExternalRefs: boolean;
  resolveReference(raw: string, from: Swagger20Resource, memo?: Swagger20ResolutionMemo): Promise<Swagger20Resolution>;
  resolvePathItem(item: Swagger20PathItem, method: Swagger20Method, memo?: Swagger20ResolutionMemo): Promise<Swagger20PathItem>;
  rebind(options: Swagger20LoadOptions): Swagger20ReferenceGraphContract;
  rememberResolvedResource(requested: string | undefined, retrieval: string | undefined, root: unknown): Swagger20Resource;
}

/** @internal */
export interface Swagger20Resolution {
  node?: unknown;
  resource: Swagger20Resource;
  cycle?: boolean;
}

/** @internal */
export interface Swagger20ResolutionMemo {
  active: Set<string>;
  pathActive: Set<string>;
  done: Map<string, Swagger20Resolution>;
}

/** Presence-aware field result. */
export interface Swagger20Member<T> {
  value?: T;
  present: boolean;
  valid: boolean;
}

/** @internal */
export function member(object: Swagger20Object, name: string): Swagger20Member<unknown> {
  if (!Object.hasOwn(object, name)) return { present: false, valid: false };
  return { value: object[name], present: true, valid: true };
}

/** @internal */
export function stringMember(object: Swagger20Object, name: string): Swagger20Member<string> {
  const found = member(object, name);
  return found.present && typeof found.value === "string"
    ? { value: found.value, present: true, valid: true }
    : { present: found.present, valid: false };
}

/** @internal */
export function booleanMember(object: Swagger20Object, name: string): Swagger20Member<boolean> {
  const found = member(object, name);
  return found.present && typeof found.value === "boolean"
    ? { value: found.value, present: true, valid: true }
    : { present: found.present, valid: false };
}

/** @internal */
export function objectMember(object: Swagger20Object, name: string): Swagger20Member<Swagger20Object> {
  const found = member(object, name);
  return found.present && isSwagger20Object(found.value)
    ? { value: found.value, present: true, valid: true }
    : { present: found.present, valid: false };
}

/** @internal */
export function arrayMember(object: Swagger20Object, name: string): Swagger20Member<unknown[]> {
  const found = member(object, name);
  return found.present && Array.isArray(found.value)
    ? { value: found.value, present: true, valid: true }
    : { present: found.present, valid: false };
}

/** @internal */
export function isSwagger20Object(value: unknown): value is Swagger20Object {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

/** @internal */
export function resourceBase(resource: Swagger20Resource | undefined): string | undefined {
  return resource?.retrieval ?? resource?.requested;
}

/** @internal */
export function operationInfo(operation: Swagger20ResolvedOperation, ref: string): Swagger20OperationInfo {
  const operationId = stringMember(operation.raw, "operationId");
  const summary = stringMember(operation.raw, "summary");
  const rawTags = arrayMember(operation.raw, "tags");
  const tags = rawTags.valid ? rawTags.value!.filter((tag): tag is string => typeof tag === "string") : [];
  return {
    ref,
    path: operation.path,
    method: operation.method,
    ...(operationId.valid ? { operationId: operationId.value } : {}),
    ...(summary.valid ? { summary: summary.value } : {}),
    tags,
  };
}

/** @internal */
export function pathItemMemberResource(item: Swagger20PathItem, name: string): Swagger20Resource {
  return item.memberResources?.get(name) ?? item.resource;
}
